### ISSUE TITLE
Withdrawals tracking issue

### ISSUE THREAD

**Issue by djrtwo** [ethereum/consensus-specs#2758]
Withdrawals (code name Capella) will be specified as three features:

1. [x] Withdrawable 0x01 validators -> withdrawn (at epoch boundaries) and core withdrawal mechanism -- #2836 
2. [x] 0x00 -> 0x01 withdrawal credentials change operation -- #2855 
3. [x] Partial withdrawal on amounts above 32 eth for 0x01 validators -- #2862 
4. [x] consider removal of `withdrawn_epoch` added in (1) #2998
     *  just utilize withdrawabl_epoch and zero balance to know if safe to re-use
6. [x] consider deposits into withdrawn validators #2998

**Comment by djrtwo** [comment]
My scratchpad notes I'm working off of
https://notes.ethereum.org/BHG6JUibQ8CNqpUeVHd1dw?edit

**Comment by hwwhww** [comment]
@djrtwo  Should we add a task item of creating an EIP to handle `WithdrawalReceipt`?

**Comment by djrtwo** [comment]
yes

**Comment by arnetheduck** [comment]
Apart from the economic stuff, do we want to do something about the size of the validator list in the beacon state before pulling the trigger on this one? Those lists are becoming quite nasty to manage, specially with the multiple linear scans (quadratic if you do it the pure spec way, but linear with the right optimizations) that we do regularly.

**Comment by Jiggnubs** [comment]
No, keep it coming.   This bitcoin thingy is the greatest current to ever
happen.  When everyone is super, no one is creator, and everyone is just
graceful.  LoL be the beta !

**Comment by djrtwo** [comment]
@arnetheduck do you mean re-using indices?
The way this is spec'd allows for re-use but does not enable re-use

Or do you mean something else entirely?

**Comment by arnetheduck** [comment]
> @arnetheduck do you mean re-using indices?

yes indeed - re-using, making the list sparse or any other approach.

**Comment by mcdee** [comment]
Is it worth attempting to optimize here before withdrawals are implemented, seems a little premature as we don't know how many exited validators there will be.  Currently there are 869 out of 431,496 validators exited, or a little over 0.2%.  I know that is the lower bound, but until we have some sort of idea of the real-world number we'd be doing a fair bit of work for what could be minimal gains.

Once capella goes live we should have some idea of the churn, which would help to inform the best way to approach reduction in size of the validator list.

**Comment by Where-2** [comment]
Lost

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by gigamesh** [comment]
Could someone point me to an ELI5 of how the withdrawal process will work for the end user? After reading the specs, its not clear to me how the priority is determined. [EIP-4895](https://eips.ethereum.org/EIPS/eip-4895) mentions the withdrawal  "has no associated gas costs", but there is a max per block. So what determines who gets to go first?

**Comment by dapplion** [comment]
> Could someone point me to an ELI5 of how the withdrawal process will work for the end user? After reading the specs, its not clear to me how the priority is determined. [EIP-4895](https://eips.ethereum.org/EIPS/eip-4895) mentions the withdrawal "has no associated gas costs", but there is a max per block. So what determines who gets to go first?

Full withdrawals: First come first serve
Partial withdrawals: Sequential order by index starting at 0

**Comment by Jiggnubs** [comment]
Whatever is specified but to avoid newbies like me, include an dummy note
walk-through would help speed the process up a bit.  To also answer what
new quest arises.  I am frank to say, pairing abilities without cost or
cause but back to basics what is pure simple.  Can some body recite the
alfabetical randomly always at any given time? Or tell my goosebumps to
wave at your command? Remember good always prevails and being bad at
everything gives perfect a sense to humor.  Thank you for quick response
dear One... don't hesitate on my weirdness, I'm just like another
conscience?
